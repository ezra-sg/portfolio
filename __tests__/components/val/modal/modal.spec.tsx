import { renderWithLanguage } from '@/__tests__/testing-helpers';

import Modal from '@/components/val/modal/modal';
import { act, fireEvent } from '@testing-library/react';


jest.mock('react-icons/md', () => ({
    MdClose: () => <div>x icon</div>,
}));

describe('Modal', () => {
    const closeDialogSpy = jest.fn();
    const openDialogSpy = jest.fn();
    const modalNode = (
        <Modal
            description="description text"
            title="title text"
            trigger={<button>trigger text</button>}
        >
            <div>children text</div>
        </Modal>
    );

    afterEach(() => {
        jest.restoreAllMocks();
    });

    beforeAll(() => {
        // address error "Not implemented: HTMLDialogElement.prototype.[close/showModal]"
        window.HTMLDialogElement.prototype.close = closeDialogSpy;
        window.HTMLDialogElement.prototype.showModal = openDialogSpy;
    });

    it('should render properly', () => {
        jest.spyOn(document, 'addEventListener');
        jest.spyOn(document, 'removeEventListener');
        const result = renderWithLanguage(modalNode);
        // addEventListener is being called with 'selectionchage' for some reason not related to the code, perhaps
        // a bug in the testing library. it needs to be reset to 0 before the assertions
        (document.addEventListener as jest.Mock).mockReset();

        expect(result.container).toMatchSnapshot();
        expect(document.addEventListener).not.toHaveBeenCalled();
        expect(document.removeEventListener).not.toHaveBeenCalled();
        expect(openDialogSpy).toHaveBeenCalledTimes(0);
        expect(closeDialogSpy).toHaveBeenCalledTimes(1); // called once on initial render in useEffect > modalIsVisible
        expect(document.body.style).toMatchObject(expect.objectContaining({
            overflow: '',
            filter: '',
        }));
        expect(document.body.getAttribute('inert')).toBeFalsy();

        const triggerElement = result.queryByTestId('modal-trigger');
        expect(triggerElement).toBeInTheDocument();

        // open the modal
        // note that in the browser environment, the HTMLElementDialog.prototype.showModal() method
        // handles the opening of the modal but it doesn't change the markup at all. Thus in the test environment,
        // there is no direct indication that the modal is open. We can, however, check the side effects
        act(() => {
            fireEvent.click(triggerElement!);
        });

        expect(openDialogSpy).toHaveBeenCalledTimes(1);
        expect(closeDialogSpy).toHaveBeenCalledTimes(1);
        expect(document.addEventListener).toHaveBeenCalledTimes(2);
        const clickawayHandler = (document.addEventListener as jest.Mock).mock.calls[0][1];
        const tabFocusTrapHandler = (document.addEventListener as jest.Mock).mock.calls[1][1];
        expect(document.addEventListener).toHaveBeenCalledWith('mousedown', clickawayHandler);
        expect(document.addEventListener).toHaveBeenCalledWith('keydown', tabFocusTrapHandler);
        expect(document.removeEventListener).not.toHaveBeenCalled();

        expect(document.body.style).toMatchObject(expect.objectContaining({
            overflow: 'hidden',
            filter: 'blur(2px)',
        }));
        expect(document.body.getAttribute('inert')).toBe('true');

        const closeButton = result.queryByTestId('modal-close-button');
        expect(closeButton).toBeInTheDocument();

        // close the modal by clicking the X icon
        act(() => {
            fireEvent.click(closeButton!);
        });

        expect(document.addEventListener).toHaveBeenCalledTimes(2); // unchanged
        expect(document.removeEventListener).toHaveBeenCalledTimes(2);
        expect(document.removeEventListener).toHaveBeenNthCalledWith(1, 'mousedown', clickawayHandler);
        expect(document.removeEventListener).toHaveBeenNthCalledWith(2, 'keydown', tabFocusTrapHandler);
        expect(openDialogSpy).toHaveBeenCalledTimes(1);
        expect(closeDialogSpy).toHaveBeenCalledTimes(2); // called again
        expect(document.body.style).toMatchObject(expect.objectContaining({
            overflow: '',
            filter: '',
        }));
        expect(document.body.getAttribute('inert')).toBeFalsy();
    });

    it('should close on clickaway', () => {
        jest.spyOn(document, 'addEventListener');

        const result = renderWithLanguage(modalNode);
        const triggerElement = result.queryByTestId('modal-trigger');
        expect(triggerElement).toBeInTheDocument();

        // open the modal
        act(() => {
            fireEvent.click(triggerElement!);
        });

        const clickawayHandler = (document.addEventListener as jest.Mock).mock.calls[0][1];

        // simulate a click outside the modal
        act(() => {
            clickawayHandler(new KeyboardEvent('mousedown'));
        });

        expect(document.body.style).toMatchObject(expect.objectContaining({
            overflow: '',
            filter: '',
        }));
        expect(document.body.getAttribute('inert')).toBeFalsy();
    });
});
