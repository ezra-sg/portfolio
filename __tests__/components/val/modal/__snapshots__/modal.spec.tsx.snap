// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render properly 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-haspopup="dialog"
    data-testid="modal-trigger"
    role="button"
    tabindex="0"
    title="description text"
  >
    <button>
      trigger text
    </button>
  </div>
  <dialog
    aria-modal="true"
    class=" w-[90vw] max-w-[1000px] min-h-[40vh] max-h-[80vh] md:max-h-[600px] fixed top-0 right-0 bottom-0 left-0 items-center justify-center backdrop:bg-stone-900 backdrop:opacity-90"
  >
    <div
      class="relative w-[90vw] max-w-[1000px] min-h-[40vh] max-h-[80vh] md:max-h-[600px] rounded-sm"
    >
      <div>
        <header
          class="sticky top-0 left-0 right-0 shadow-sm flex justify-between items-center bg-amber-50 dark:bg-stone-900"
        >
          <div
            class="ml-4 my-2 dark:text-amber-50"
          >
            <h1
              class="font-header"
            >
              title text
            </h1>
            <p
              class="text-xs italic"
            >
              global.audio_transcript
            </p>
          </div>
          <button
            aria-label="Close modal for: title text"
            class="h-8 w-8 m-4 flex items-center justify-center rounded-full border-[1px] border-amber-900 hover:border-[2px] dark:border-amber-200"
            data-testid="modal-close-button"
            title="Close modal for: title text"
          >
            <div>
              x icon
            </div>
          </button>
        </header>
        <div
          class="p-4 dark:bg-slate-950 dark:text-amber-50"
        >
          <div>
            children text
          </div>
        </div>
      </div>
    </div>
  </dialog>
</div>
`;
